src/front/glsl/ast.rs

```diff
@@ -1,142 +1,150 @@
-    /// represented as `all(equal(vec1, vec2))` and `any(notEqual(vec1, vec2))`
-    pub fn equality_expr(
-        &mut self,
-        equals: bool,
-        left: &ExpressionRule,
-        right: &ExpressionRule,
-    ) -> Result<ExpressionRule, ErrorKind> {
-        let left_is_vector = match *self.resolve_type(left.expression)? {
-            crate::TypeInner::Vector { .. } => true,
-            _ => false,
-        };
+            input_struct: module.types.append(Type {
+                name: None,
+                inner: TypeInner::Struct {
+                    level: crate::StructLevel::Root,
+                    members: Vec::new(),
+                    span: 0,
+                },
+            }),
+            output_struct: module.types.append(Type {
+                name: None,
+                inner: TypeInner::Struct {
+                    level: crate::StructLevel::Root,
+                    members: Vec::new(),
+                    span: 0,
+                },
+            }),
 
-        let right_is_vector = match *self.resolve_type(right.expression)? {
-            crate::TypeInner::Vector { .. } => true,
-            _ => false,
-        };
+            module,
+        }
+    }
 
-        let (op, fun) = match equals {
-            true => (BinaryOperator::Equal, RelationalFunction::All),
-            false => (BinaryOperator::NotEqual, RelationalFunction::Any),
+    pub fn resolve_type<'b>(
+        &'b mut self,
+        context: &'b mut Context,
+        handle: Handle<Expression>,
+    ) -> Result<&'b TypeInner, ErrorKind> {
+        let resolve_ctx = ResolveContext {
+            constants: &self.module.constants,
+            global_vars: &self.module.global_variables,
+            local_vars: &context.locals,
+            functions: &self.module.functions,
+            arguments: &context.arguments,
         };
-
-        let expr =
-            ExpressionRule::from_expression(self.context.expressions.append(Expression::Binary {
-                op,
-                left: left.expression,
-                right: right.expression,
-            }));
-
-        Ok(if left_is_vector && right_is_vector {
-            ExpressionRule::from_expression(self.context.expressions.append(
-                Expression::Relational {
-                    fun,
-                    argument: expr.expression,
-                },
-            ))
-        } else {
-            expr
-        })
+        match context.typifier.grow(
+            handle,
+            &context.expressions,
+            &mut self.module.types,
+            &resolve_ctx,
+        ) {
+            //TODO: better error report
+            Err(error) => Err(ErrorKind::SemanticError(
+                format!("Can't resolve type: {:?}", error).into(),
+            )),
+            Ok(()) => Ok(context.typifier.get(handle, &self.module.types)),
+        }
     }
 
-    pub fn resolve_type(
+    pub fn resolve_handle(
```

Comment:
We shouldn't need to have this as a separate method. Instead, we could add `fn get_resolution()` method to `Typifier`, and return `TypeResolution` here in this method. Then everything else can be done on it:
  - if you need a `TypeInner`, then call `resolution.inner_with(types)`
  - if you need `Handle<Type>`, then call `resolution.handle().unwrap()`

Reply:
hmm I'm not sure that would work since the frontend would panic on things like float literals.

Ideally I would like to not have this at all because right now this is used in `FunctionSignature` to support overloading but this is really error prone because we must ensure that all the types that are equal must have the same handle.